eventmesh/ApiServerLockManager.js

```diff
@@ -0,0 +1,179 @@
+              }
+          })    
+          }
+      }
+  }
+  Lock is tracked via resource of resource group lock and resource type deploymentlock.
+  id of the resource is instance_id.
+  To lock the deployment we create a new resource of resourceType deploymentlock.
+  If the resource is already present then deployment can't be locked.
+  Lock Algorithm looks like the following
+      1. Create resource lock/deploymentlock/instance_id
+      2. If resource is already present than can't acquire lock
+      3. Check lockdetails from spec.options, if TTL is expired than update the lockdetails
+      4. else can't acquire lock
+  */
+
+  /**
+   * @param {string} resourceId - Id (name) of the resource that is being locked. In case of deployment lock, it is instance_id
+   * @param {object} lockDetails - Details of the lock that is to be acquired
+   * @param {number} [lockDetails.lockTTL=Infinity] - TTL in miliseconds for the lock that is to be acquired
+   * @param {string} lockDetails.lockType - Type of lock ('READ'/'WRITE')
+   * @param {object} [lockDetails.lockedResourceDetails] - Details of the operation who is trying to acquire the lock
+   * @param {string} [lockDetails.lockedResourceDetails.resourceType] - Type of resource for which lock is being acquired. ex: backup
+   * @param {string} [lockDetails.lockedResourceDetails.resourceName] - Name of resource for which lock is being acquired. ex: defaultbackup
+   * @param {string} [lockDetails.lockedResourceDetails.resourceId] - Id of resource for which lock is being acquired. ex: <backup_guid>
+   * @param {string} [lockDetails.lockedResourceDetails.operation=unknown] - Operation type who is acquiring the lock. ex: backup
+   */
+
+  lock(resourceId, lockDetails) {
+    if (!lockDetails) {
+      lockDetails = {};
+    }
+    const currentTime = new Date();
+    const opts = _.cloneDeep(lockDetails);
+    opts.lockTTL = opts.lockTTL ? opts.lockTTL : Infinity;
+    _.extend(opts, {
+      'lockTime': currentTime
+    });
+    logger.info(`Attempting to acquire lock on resource with resourceId: ${resourceId} `);
+    return eventmesh.server.getResource(CONST.APISERVER.RESOURCE_TYPES.LOCK, CONST.APISERVER.RESOURCE_NAMES.DEPLOYMENT_LOCKS, resourceId)
+      .then(resource => {
+        const resourceBody = resource.body;
+        const currentlLockDetails = JSON.parse(resourceBody.spec.options);
+        const currentLockTTL = currentlLockDetails.lockTTL ? currentlLockDetails.lockTTL : Infinity;
+        const currentLockTime = new Date(currentlLockDetails.lockTime);
+        if ((currentTime - currentLockTime) < currentLockTTL) {
+          logger.error(`Resource ${resourceId} was locked for ${currentlLockDetails.lockedResourceDetails.operation} ` +
+            `operation with id ${currentlLockDetails.lockedResourceDetails.resourceId} at ${currentLockTime} `);
+          throw new EtcdLockError(`Resource ${resourceId} was locked for ${currentlLockDetails.lockedResourceDetails.operation} ` +
```

Comment:
I would prefer renaming this to `ResourceLockError` since the broker framework changes need not be aware of Etcd as the backend for API Server at all. Also `EtcdLockError` clashes with the error class from the `etcd3` module as well